src/ComponentBase.ts

```diff
@@ -117,12 +123,13 @@ abstract class ComponentBase<P extends React.Props<any>, S extends Object> exten
         });
 
         // Initialize state
-        this.state = (this._buildStateWithAutoSubscriptions(this.props, true) as S) || ({} as S);
+        this._nextState = (this._buildStateWithAutoSubscriptions(this.props, true) as S) || ({} as S);
+        this.setState(this._nextState);
         this._isMounted = true;
     }
 
     // Subclasses may override, but _MUST_ call super.
-    componentWillReceiveProps(nextProps: Readonly<P>, nextContext: any): void {
+    componentWillReceiveProps<K extends keyof S>(nextProps: Readonly<P>, nextContext: any): void {
```

Comment:
Remove the type parameter. It is not giving you safety: it is either inferred as some default type (like `string` or `any`) or explicitly specified by the caller which is not actually helpful. There is nothing in the compilation that connects the call site of `componentWillReceiveProps<K>` with how `K` is used below, so the compiler is not checking anything.

Reply:
I agree it doesn't add type safety.

This is the only way to pass something that looks like a correct type I found.

Do you suggest to use any in the method body instead?